### ISSUE TITLE
Error in concretization when dependency adds constraint

### ISSUE THREAD

**Issue by nazavode** [spack/spack#7533]
Given the following packages (boilerplate omitted):

```python
class TestDependent(Package):
    ...
    depends_on('boost')
    depends_on('test-dependency')
    ...
```

```python
class TestDependency(Package):
    ...
    depends_on('boost           +system +filesystem', when='@:1.1.999')
    depends_on('boost@:1.65.999 +system +filesystem', when='@1.2:')
    ...
```

The `spec` concretization fails with a conflict on `boost`:

```console
$ spack spec test-dependent
Input spec
--------------------------------
     test-dependent

Concretized
--------------------------------
==> Error: An unsatisfiable version constraint has been detected for spec:

    boost@1.66.0%clang@9.0.0-apple+atomic+chrono~clanglibcpp+date_time~debug+exception+filesystem~graph~icu+iostreams+locale+log+math~mpi+multithreaded+program_options~python+random+regex+serialization+shared+signals~singlethreaded+system~taggedlayout+test+thread+timer~versionedlayout+wave arch=darwin-highsierra-x86_64
        ^bzip2
        ^zlib


while trying to concretize the partial spec:

    test-dependency@2.0%clang@9.0.0-apple arch=darwin-highsierra-x86_64


test-dependency requires boost version :1.65.999, but spec asked for 1.66.0
```
This looks weird given the fact that the dependent package poses no constraints on `boost` whatsoever.

When a feasible version of `boost` is pinned right in the root *abstract* spec, everything works as expected:

```console
$ spack spec test-dependent^boost@1.65.1
Input spec
--------------------------------
test-dependent
    ^boost@1.65.1

Concretized
--------------------------------
test-dependent@1.0%clang@9.0.0-apple arch=darwin-highsierra-x86_64
    ^boost@1.65.1%clang@9.0.0-apple+atomic+chrono~clanglibcpp+date_time~debug+exception+filesystem~graph~icu+iostreams+locale+log+math~mpi+multithreaded patches=2ab6c72d03dec6a4ae20220a9dfd5c8c572c5294252155b85c6874d97c323199 +program_options~python+random+regex+serialization+shared+signals~singlethreaded+system~taggedlayout+test+thread+timer~versionedlayout+wave arch=darwin-highsierra-x86_64
        ^bzip2@1.0.6%clang@9.0.0-apple+shared arch=darwin-highsierra-x86_64
        ^zlib@1.2.11%clang@9.0.0-apple+optimize+pic+shared arch=darwin-highsierra-x86_64
    ^test-dependency@2.0%clang@9.0.0-apple arch=darwin-highsierra-x86_64
```

The same goes when the dependency is pinned to a version that considers the latest `boost` as feasible:

```console
$ spack spec test-dependent^test-dependency@1.1
Input spec
--------------------------------
test-dependent
    ^test-dependency@1.1

Concretized
--------------------------------
test-dependent@1.0%clang@9.0.0-apple arch=darwin-highsierra-x86_64
    ^boost@1.66.0%clang@9.0.0-apple+atomic+chrono~clanglibcpp+date_time~debug+exception+filesystem~graph~icu+iostreams+locale+log+math~mpi+multithreaded patches=2ab6c72d03dec6a4ae20220a9dfd5c8c572c5294252155b85c6874d97c323199 +program_options~python+random+regex+serialization+shared+signals~singlethreaded+system~taggedlayout+test+thread+timer~versionedlayout+wave arch=darwin-highsierra-x86_64
        ^bzip2@1.0.6%clang@9.0.0-apple+shared arch=darwin-highsierra-x86_64
        ^zlib@1.2.11%clang@9.0.0-apple+optimize+pic+shared arch=darwin-highsierra-x86_64
    ^test-dependency@1.1%clang@9.0.0-apple arch=darwin-highsierra-x86_64
```

### Expected Result

To have the spec tree concretized with a `boost` version that satisfies the constraint posed by the leaf dependency, something like:

```console
$ spack spec test-dependent
Input spec
--------------------------------
test-dependent

Concretized
--------------------------------
test-dependent@1.0%clang@9.0.0-apple arch=darwin-highsierra-x86_64
    ^boost@1.65.1%clang@9.0.0-apple+atomic+chrono~clanglibcpp+date_time~debug+exception+filesystem~graph~icu+iostreams+locale+log+math~mpi+multithreaded patches=2ab6c72d03dec6a4ae20220a9dfd5c8c572c5294252155b85c6874d97c323199 +program_options~python+random+regex+serialization+shared+signals~singlethreaded+system~taggedlayout+test+thread+timer~versionedlayout+wave arch=darwin-highsierra-x86_64
        ^bzip2@1.0.6%clang@9.0.0-apple+shared arch=darwin-highsierra-x86_64
        ^zlib@1.2.11%clang@9.0.0-apple+optimize+pic+shared arch=darwin-highsierra-x86_64
    ^test-dependency@1.2%clang@9.0.0-apple arch=darwin-highsierra-x86_64
```

### Actual Result
Concretization fails with a conflict:

```console
Error: An unsatisfiable version constraint has been detected for spec
```

### Steps to reproduce the issue

I've set up a [minimal Spack repo](https://github.com/nazavode/spack-repo-test.git) with two bogus packages (`test-dependent` and `test-dependency` as shown above) able to reproduce the issue:

```console
$ git clone https://github.com/nazavode/spack-repo-test.git
$ spack repo add spack-repo-test
$ spack spec test-dependent
...
test-dependency requires boost version :1.65.999, but spec asked for 1.66.0
```

### Information on your system

Discovered on `arch=darwin-highsierra-x86_64` and `Python 2.7.10` but the issue seems to be reproducible regardless of `arch` and/or python version.

**Comment by alalazo** [comment]
Duplicate of #1781

**Comment by alalazo** [comment]
The issue here is the fact that with the original concretizer unconditional dependencies are evaluated first. If a later conditional dependency constrain them further the concretizer fails since it doesn't backtrack. Solved by #19501
